README.md
PROJECT CONTENTS:
--------------
:heart: :deciduous_tree: :construction_worker: **Arterial Tree Builder** (ATB) is a C++ project that outputs InteropFFR.dll used by
MultiVox software. ATB library contains semi-automatic algorithms for:
- [IDT](https://link.springer.com/chapter/10.1007/11744078_35)-based aorta segmentation
- [Frangi Vesselness](https://link.springer.com/chapter/10.1007/BFb0056195) based coronary and brachiocephalic arteries
- Voxel skeletonization based on [DOHT](https://www.sciencedirect.com/science/article/pii/S1077314298906804)
- Skeleton-based vessel 1D graph construction
- Pathology marker setting on 1D graph

All user interation is provided in MultiVox GUI.

**Tortuosity Prototype** is a python-based utility to smooth 3D lines, compute curvature in every point, and draw the results. 
Inputs voxel centerlines of cerebral vessels.

REQUIREMENTS LIST:
--------------
0. Works only for Windows
1. Visual Studio 2015.
2. [Cmake-GUI](https://github.com/Kitware/CMake/releases/download/v3.15.0/cmake-3.15.0-win64-x64.msi).
3. ITK library (ver 5+) + Win Git.
4. MultiVox (graphical interface that exploites ATB output dll).
5. GraphViz (extra package for arterial graph visualization).
6. Open MP, OpenCl Runtime


BUILDING AND INSTALLING ITK LIBRARY
--------------
1. Install [Git for Windows](https://git-scm.com/download/win).
2. Download ITK 5.0.1 (or higher) [source](https://itk.org/ITK/resources/software.html)
3. Open Cmake-GUI, specify ITK source folder and build folder.
4. Press "Configure" button and set generator to "Visual Studio 14 2015 Win64".
5. Change CMAKE_INSTALL_PREFIX not to be inside system folder, 
e.g. D:/libraries/ITK (not C:/Program Files/ITK !)
Change CMAKE_CONFIGURATION_TYPES to required build type
6. Set checkbox "Module_IOMeshSTL" ON (this step requires Git)
7. Press "Configure" again, and then "Generate".
8. Open builded project with Visual Studio, right-click on "ALL BUILD" and build it.
9. To install ITK to specified path, right-click on "INSTALL" and build it.

*. [Alternative installation guide](https://itk.org/Wiki/ITK/Configuring_and_Building/VisualStudio)



BUILDING ATK
------------
1. Open ATK project with Visual Studio.
2. Go to Project->Project Properties. Specify next fields:

For Release build:\\
* "отладка" -> "Команда" = path_to_MultiVox\Bin\MultiVox.exe
* "отладка" -> "Рабочий каталог" = path_to_MultiVox\Bin
* "C/C++" -> "Дополнительные каталоги включаемых файлов" = ITK_RELEASE_install_path\include\ITK-4.12\;$(VC_IncludePath)headers;
* "C/C++" -> "Язык" -> "Поддержка Open MP"  = Да(/openmp)
* "C/C++" -> "Создание кода" -> "Runtime library"  = Многопоточный DLL (/MD)
* "C/C++" -> "Оптимизация" -> "Оптимизация" = Максимальная скорость (/O2)
* "C/C++" -> "Оптимизация" -> "Оптимизация всей программы" = Да (/GL)
* "Linker" -> "Общие"->"Дополнительные каталоги библиотек" = ITK_RELEASE_install_path\lib;%(AdditionalLibraryDirectories)
* "Linker" -> "Ввод" -> "Дополнительные зависимости" -> ITK_RELEASE_install_path\lib\*.lib;%(AdditionalDependencies)
* ????? "Linker" -> "Ввод" -> "Файл Определения Модуля" = .\InteropFFR3.1.def
* "События сборки" -> "События после сборки" -> "Командная строка" -> copy "$(TargetDir)\InteropFFR.dll" "path_to_MultiVox\Bin"
* "Ресурсы" -> "Общие" -> NDEBUG;%(PreprocessorDefinitions)

For Debug build the next lines are different:
* "C/C++" -> "Дополнительные каталоги включаемых файлов" = ITK_DEBUG_install_path\include\ITK-4.12\;$(VC_IncludePath)headers;
* "C/C++" -> "Оптимизация" -> "Оптимизация" = Отключено (/Od)
* "C/C++" -> "Оптимизация" -> "Оптимизация всей программы" = Нет
* "C/C++" -> "Создание кода" -> "Runtime library"  = Многопоточная отладка DLL (/MDd)
* "Linker" -> "Общие"->"Дополнительные каталоги библиотек" = ITK_DEBUG_install_path\lib;%(AdditionalLibraryDirectories)
* "Linker" -> "Ввод" -> "Дополнительные зависимости" -> ITK_DEBUG_install_path\lib\*.lib;%(AdditionalDependencies)
* "Ресурсы" -> "Общие" -> _DEBUG;%(PreprocessorDefinitions)


INSTALLING MultiVox
------------------
//TODO: MultiVox Installation Guide
1. Install x86 runtime by rinnning SSCERuntime_x86-RUS.msi
2. Install x64 runtime by rinnning SSCERuntime_x64-RUS.msi
3. Install Multivox by running multivox.msi. MultiVox will be installed to standart folder C:\Program Files (x86)\Alda Universal\Multivox
4. Download and install OpenCL runtime to use your CPU with MultiVox. Donwload with registration [here](https://software.intel.com/en-us/articles/opencl-drivers) or find installer inside of "MultiVox_files".
5. Copy and replace InteropFFR.dll, MultiVox.exe.config from "MultiVox_files\" to "C:\Program Files (x86)\MultiVox\Bin".
6. Run TaskManager(Ctrl+Shift+Esc) and stop process MultiVox.ClientService.exe and run Multivox\Bin\MutiVox.config.exe, go to "Units", click on line "3D-EVA Mode", and push preference button on the right (with the spanner picture). In "OpenCL" tab, specify your GPU device and CPU device. If you do not have GPU on your computer, CPU device will be used with OpenCL runtime. Click Apply, OK, and OK (resturt client services MultiVox). In TaskManager start MultiVox.ClientServices again.

After this step run C:\Program Files (x86)\Alda Universal\MultiVox\Bin\MultiVox.exe , load data (File->read Dicom Folder)  and check that tab "Instruments" has field "Initialize FFR preprocessing".

Note that C:\Program Files(x86)\Alda Universal\MultiVox is the system folder, so ATB will not be able to write logs and other files inside of it.

7*. Copy this folder to your local folder to ged rid of this issue.


PREPARING FFR MODULE
--------
1. Copy FFR folder to MultiVox\Bin\
2. Copy Alda.MultiVox.Vessels.FFR.Base.FFRSettings.xml from "MultiVox_files\(net)_MultiVox 3D (DEMO,x64,NET40)_6.1.7033.47296\Program Files\Alda Universal\MultiVox\Bin" to MultiVox\Bin\
3. Open MultiVox\Bin\Alda.MultiVox.Vessels.FFR.Base.FFRSettings.xml and specify absolute path to FFR\blood.exe

INSTALLING GraphViz
--------------------
1. Download [GraphViz](https://graphviz.gitlab.io/_pages/Download/Download_windows.html) msi-installer
2. Run installer
3. Bind C:\Program Files (x86)\Graphviz2.38\bin\gvedit.exe to *.gv files



PAPERS
--------
1. [Methods of graph network reconstruction in personalized medicine](https://dodo.inm.ras.ru/research/_media/cnm2754.pdf)
2. [Image Segmentation for Cardiovascular Biomedical Applications at Different Scales](https://pdfs.semanticscholar.org/ced1/7f04d5a58e9fbc04847d935f9bd5b51065d4.pdf)
3. [Robustness Analysis of Coronary Artery Segmentation](https://link.springer.com/chapter/10.1007/978-3-030-06228-6_26)
4. [Patient-specific anatomical models in human physiology](https://dodo.inm.ras.ru/research/_media/rjnamm_vassilevski15.pdf)


Changelog.txt
19.10.22 <<< 1.2.4.72 новый формат FFRfull.tre >>> ffr-min.js 262 688 warn:102
17.10.22 <<< 1.2.4.72 ЧСС и другие данные пациента >>> ffr-min.js 262 647 warn:102
15.10.22 <<< 1.2.4.71 >>> ffr-min.js 260 623 warn:101
 * колесико, PAN и ZOOM
 * табличка с ФРК
26.09.22 <<< 071RC1 табличка с ФРК>>>ffr-min.js 282 578 warn:99
20.08.22 <<< 071RC колесико, PAN и ZOOM >>>ffr-min.js 281 298 warn:97
19.08.22 <<< 1.2.4.70 FindAorta со многими срезами >>>ffr-min.js 282 774 warn:96
13.08.22 <<< 070RC FindAorta со многими срезами >>>ffr-min.js 282 218 warn:96
09.08.22 <<< 070RC RGT: первые регрессионные тесты >>>ffr-min.js 281 847 warn:96
07.08.22 <<< 1.2.4.69 Удалить сосуд >>>ffr-min.js 275 937 warn:95
05.08.22 <<< 1.2.4.68 ускорена сегментация аорты >>>ffr-min.js 274 704 warn:94
 * ускорена сегментация аорты (класс IDT) [Данилов]
 * более информативные проценты в прогрессбаре [Данилов]
 * cнижена вероятность обрыва связи [Данилов]
 * Отображение сглаженной аорты при выборе порога сегментации коронаров [Данилов]

05.08.22 <<< более информативные проценты в прогрессбаре >>>
04.08.22 <<< 1.2.4.67 ROI с рамками >>>ffr-min.js 274 331 warn:94
 * появился интерфейс по управлению рамками ROI
 * тест план при выпуске релиза
 * bugfix: ошибка при Закрыть серию (null rhs)
03.08.22 <<< ускорена сегментация аорты (класс IDT) >>>
30.07.22 <<< 1.2.4.66 сборка с GetOptimalTau >>> ffr-min.js 266 084 warn:86
* интегрирован pull request 1 Саши Данилова с автоматическим определением порого tau

28.07.22 <<< 1.2.4.66 добавлен GetOptimalTau в протокол >>> ffr-min.js 266 072 warn:86
18.07.22 <<< 1.2.4.65 series resize и докерфайлы для CRADLE ATB >>> ffr-min.js 265 985 warn:86
13.07.22 <<< 1.2.4.65 https и wss в PROD >>> ffr-min.js 264 479 warn:81
07.06.22 <<< 1.2.4.64 с новым blood, LUT, state4 Брусницин, свет >>> ffr-min.js 264 479 warn:81
26.05.22 <<< 1.2.4.63 учет пустых воксельных масок при передаче >>> ffr-min.js 263 795 warn:81
25.05.22 <<< 1.2.4.62 PlasticBoy с LUT >>> ffr-min.js 263 542 warn:82
21.05.22 <<< 1.2.4.61 FFRserver c логом OpenMP threads >>> ffr-min.js 260 190 warn:86
19.05.22 <<< скрипт collect:PROD >>>
17.05.22 <<< 1.2.4.60 FFRserver с включенным OpenMP >>> ffr-min.js 258 240 warn:82
15.05.22 <<< 1.2.4.59 lws ping disabled; полная функциональность маркировки>>> ffr-min.js 258 240 warn:82
10.05.22 <<< 1.2.3.58 lws ping>>> ffr-min.js 255 287 warn:82
07.05.22 <<< 1.2.3.57 инфо стеноз и colored boxes>>> ffr-min.js 255 227 warn:82
07.05.22 <<< infoDiv и чистка кода >>> ffr-min.js 254 161 warn:82
02.05.22 <<< 1.2.2.56 исправлен парсинг PrepareResult >>> ffr-min.js 254 514 warn:82
01.05.22 <<< чистка кода >>> ffr-min.js 254 569 warn:82
01.05.22 <<< 1.2.2.55 Почукаева FFRfull демо >>> ffr-min.js 254 569 warn:85
30.04.22 <<< colorVesselSegment перенесен в VesselHistogram >>> ffr-min.js 250 574 warn:83
30.04.22 <<< makeTubeMesh перенесен из plastic boy  в Tubes >>> ffr-min.js 253 880 warn:85
29.04.22 <<< старый код по раскраске сосоудов перенесен из rainbow в plastic boy >>> 
29.04.22 <<< удаление старых selectVesselSegment(s) >>> 
29.04.22 <<< перекомпоновка selectVesselSegment(s) >>> 
29.04.22 <<< 1.2.2.54 plastic boy >>> ffr-min.js 251 953 warn:88
18.04.22 <<< 1.2.2.53 Capture/Start/Clear, UserStenosisPercent и массив стенозов в PrepareResult >>> ffr-min.js 250 385 warn:74 (только компиляция)
17.04.22 <<< 1.2.1.52 LUT и rainbow >>> ffr-min.js 248 159 warn:79
12.04.22 <<< рефакторинг-2 MarkStenosisComponent >>>
12.04.22 <<< рефакторинг-1 MarkStenosisComponent >>>
12.04.22 <<< фиксация кода перед рефакторингом MarkStenosisComponent >>>
09.04.22 <<< 1.2.1.51 работает переключение сосудов >>> ffr-min.js 255 274 warn:59
05.04.22 <<< select/deselect столбика на гистограмме >>> ffr-min.js 254 197 warn:58
05.04.22 <<< 1.2.0.50 убраны axplane из интефейса >>> ffr-min.js 254 019 warn:58
03.04.22 <<< tooltips на гистограмме >>> ffr-min.js 254 897 warn:58
03.04.22 <<< colorVesselSegment >>> ffr-min.js 254 548 warn:58
02.04.22 <<< 1.2.0.49 прототип раскраски части сосуда по клику на гистограмму >>> ffr-min.js 254 156 warn:58
30.03.22 <<< 1.2.0.48 VesselHistogramComponent >>> ffr-min.js 253 883 warn:58
23.03.22 <<< 1.1.1.47 увеличение буфера до 10К и раскраска Почукаевой 90% >>> 
11.03.22 <<< 1.1.1.46 коррекная работа Прорядить Серию >>> ffr-min.js 251 154 warn:58 (LUT)
09.03.22 <<< 1.1.0.45 новый blood - раскраска артерий значением ФРК >>> ffr-min.js 251 115 warn:58 (LUT)
03.03.22 <<< 1.1.0.44 профиль Почукаевой через Onion >>> ffr-min.js 251 490 warn:40
01.03.22 <<< 1.1.0.44 класс Onion для автоматической маркировки стеноза >>> ffr-min.js 250 903 warn:39
27.02.22 <<< 1.0.1.43 эксперименты с inflateN6 >>> ffr-min.js 250 654 warn:35
16.02.22 <<< 1.0.1.42 дополнительное освещение >>> ffr-min.js 248 485 warn:35
13.02.22 <<< 1.0.1.41 профиль Почукаевой RV3 с учетом близости к центральной линии >>> ffr-min.js 247 990 warn:35
12.02.22 <<< 1.0.1.40 получил профиль сечений на Почукаевой right vesssel 3 >>> ffr-min.js 249 185 warn:35
19.12.21 <<< 1.0.1.39 эксперименты по автоматическому нахождению среза аорты >>>
19.12.21 <<< 1.0.1.39 добавил ui/dialogs/Dialog11.js >>>
11.12.21 <<< 1.0.1.39 issue#9 (диапазон аорты) AortaSearch_FindAorta stats на Домаскине >>>
08.12.21 <<< 1.0.1.39 ui/dialogs/Dialog11.js с обработкой input enter >>> ffr-min.js 248 093 warn:35
06.12.21 <<< 1.0.0.38 базовый вариант для полевых испытаний  >>> ffr-min.js 248 074 warn:36
05.12.21 <<< issue#6.2 enable/disable OK and Cancel buttons in Dialog-10  >>> ffr-min.js 248 050 warn:37

www/js/main.js
goog.require('goog.object');
goog.require('goog.dom');
goog.require('goog.ui.Toolbar');
goog.require('goog.ui.Menu');
goog.require('goog.ui.MenuItem');
goog.require('goog.ui.SubMenu');
goog.require('goog.ui.MenuSeparator');
goog.require('goog.ui.ToolbarMenuButton');
goog.require('goog.ui.ToolbarButton');
goog.require('goog.ui.Component');
goog.require('goog.events');
goog.require('goog.events.Event');
goog.require('goog.style');

goog.require('gfk.ui.Dialog');
goog.require('goog.ui.SplitPane');
goog.require('gfk.ui.SplitPane');
goog.require('gfk.ui.SplitPaneExpanded');
goog.require('gfk.ui.TabBar');
goog.require('gfk.ui.ProgressBar');

/* Пространство имен приложения FFR
 */
var FFR = FFR || {};
FFR.ui = {};
FFR.ui.events = {};

FFR.viewer = null;
// export
window['FFR'] = FFR;

FFR.createViewer = function (parent) {
    //const date = new Date(); console.log(date.toLocaleString());    // 07.10.2021, 15:46:53
    document.title = `V5-FFR Viewer build ${AppConfig.build}`;
    var mainwin = goog.dom.createDom("div",{"class": "ffr-viewer-main"});
    parent.appendChild(mainwin);
    FFR.viewer = new Viewer(mainwin);
}
FFR['createViewer'] = FFR.createViewer;

class MyComponentResized extends goog.ui.Component {
    constructor() {
        super();
    }
    resize() {
        // console.log('MyComponentResized::resize()');
        this.forEachChild(child => child.resize());
    }
}
/**
 * Класс FFR.Viewer
 * constructor onResize onDoubleClickR0 onRightSplitpaneHandleDrug
 * createToolbar
 * processToolbarAction_
 * loadAortaMask(file) file={File|Blob} - вызывается в ffrdriver.js
 * showInfoMessage(message)
 */
class Viewer {

    constructor(mainwin) {
        console.log("constructor Viewer");
        // запрашиваем автоматичекий запуск сервиса ATB. 

        // Подключение будет позже
        //
        this.serviceRequest();

        let self = this;
        this._about = `Версия ${AppConfig.version}.${AppConfig.build}\nот ${AppConfig.date} `;

        this.mainwin = mainwin;  
        this._wsurl = AppConfig.wsurl; //'ws://localhost:8889';  
        //this._wsurl_remote = 'ws://dodo.inm.ras.ru/websocket';
        this._semafor = null;
        this._menuItemMarkStenosis = null;
        this.toolbar = this.createToolbar();
        this.markStenosis = null;

        // это теперь управляется {open|close}Series
        this._FFRsrv        = null;   
        this._ffrDriver     = null;      

        var splitmain = goog.dom.createDom("div", {});
        var container = goog.dom.createDom("div", {"class": "container", "id": "container"});
        var stage = goog.dom.createDom("div",{"class": "ffr-viewer-stage"}, container);
        this.mainwin.appendChild(stage);

        //new goog.ui.Button('hello'); 
        this._quadview = new QuadView(this);
        this.leftPanel = new LeftPanelComponent(this);
        let btmPlaceholderComponent = new MyComponentResized() //new goog.ui.Component();//new SeriesListsComponent();
        this.seriesLists = new SeriesListsComponent(); //  new VesselHistogramComponent();
        // gfk.ui.SplitPaneExpanded goog.ui.SplitPane
        this._vspr = new gfk.ui.SplitPaneExpanded(this._quadview,btmPlaceholderComponent,goog.ui.SplitPane.Orientation.VERTICAL)
        // btmPlaceholderComponent.getElement().style.height = "99%";
        btmPlaceholderComponent.addChild(this.seriesLists,true);
        // new goog.ui.Component()   new MyComponentResized()
        this._mainSplitControl = new goog.ui.SplitPane(new MyComponentResized(),this._vspr, goog.ui.SplitPane.Orientation.HORIZONTAL);
        this._mainSplitControl.getChildAt(0).addChild(this.leftPanel,true);
        this._mainSplitControl.render(container);
        this._mainSplitControl.getChildAt(0).getElement().style.height = '100%';    // это важно
        this._mainSplitControl.setFirstComponentSize(350);  // на глаз
        this._vspr.setFirstComponentSize(450);

        this.markStenosisComponent = null
        this.vesselHistogramComponent = null

        // goog.events.listen(vspr.splitpane, goog.ui.Component.EventType.CHANGE, e => this.onRightSplitpaneHandleDrug(e)); 

        window.addEventListener('resize',()=>this.onResize());

    }

    serviceRequest() {
        var xhr = new XMLHttpRequest;
        xhr.open("GET", "service.php?c=start",false);   // sync
        xhr.onload = function() {
            console.log('service.php onload()');
            console.log('получили '+xhr.response.length+' байт');
            //console.log(xhr.response);
        }
        xhr.onerror = function() {
            console.log('service.php onerror()');
        }

        console.log('sending service.php request start service')

        xhr.send();
    }

    processMarkStenosis(state) {
        // скрываем левую панель
        this._mainSplitControl.getChildAt(0).removeChild(this.leftPanel,true)   // альтернатива setVisible(false)

        // устанавливаем вместо нее MarkStenosisComponent
        let MarkStenosisComponentRef = new MarkStenosisComponent(null /*lhsSize*/, state);
        this.markStenosisComponent = MarkStenosisComponentRef;
        this._mainSplitControl.getChildAt(0).addChild(MarkStenosisComponentRef,true)

        // VesselHistogramComponent на месте SeriesLists
        //
        this.seriesLists.setVisible(false);
        this._vspr.getChildAt(1).removeChildAt(0);
        this.vesselHistogramComponent = new VesselHistogramComponent(MarkStenosisComponentRef);
        this._vspr.getChildAt(1).addChild(this.vesselHistogramComponent,true);

    }

    onResize() {
      let stage = document.getElementsByClassName('ffr-viewer-stage')[0];
      // logElementSize(stage)
      let size = goog.style.getSize(stage);
      // console.log('onResize()',size);
      this._mainSplitControl.setSize(size);
      // this._vspr.fitParent();        // это из прошлой жизни
    }

    onDoubleClickR0(e) {
      console.log('onDoubleClickR0()',e)
    }


    /**
     * FFR.Viewer.onRightSplitpaneHandleDrug()   - обработчик handle правой splitpane
     * return: goog.ui.Toolbar
     */
    onRightSplitpaneHandleDrug(e) {
      console.log('onRightSplitpaneHandleDrug()'/*, e.target*/);
    }

	/**
     * FFR.Viewer.createToolbar()
     * return: goog.ui.Toolbar
     */
    createToolbar()  {

        var t1 = new goog.ui.Toolbar();
        var self = this;

        // добавляем menuButton File
        var fileMenu = new goog.ui.Menu();
        let inputIdFile = t1.makeId('inpf'); 
        let ee =[
            goog.dom.createDom(goog.dom.TagName.INPUT, {
                id: inputIdFile,
                type: goog.dom.InputType.FILE,
                'class': 'inputfile-invisible',
              }),
            goog.dom.createDom(goog.dom.TagName.LABEL, {for:inputIdFile}, 'Чтение DICOM файла')];
        fileMenu.addItem(new goog.ui.MenuItem(ee));

        let inputId = t1.makeId('inpd'); 
        ee =[goog.dom.createDom("div","goog-menuitem-icon","📂"),
            goog.dom.createDom(goog.dom.TagName.INPUT, {
                id: inputId,
                type: goog.dom.InputType.FILE,
                'class': 'inputfile-invisible',
                'webkitdirectory':' ', 
              }),
            goog.dom.createDom("div","goog-menuitem-accel","Ctrl+F2"),
            goog.dom.createDom(goog.dom.TagName.LABEL, {for:inputId}, 'Чтение DICOM каталога')];
        fileMenu.addItem(new goog.ui.MenuItem(ee));

        var btnFile = new goog.ui.ToolbarMenuButton('Файл',fileMenu);
        btnFile.setTooltip('Тесты диалогов');
        t1.addChild(btnFile,true);

        // добавляем в меню Инструменты
        var toolsMenu = new goog.ui.Menu();
        toolsMenu.addItem(this._menuItemMarkStenosis = new goog.ui.MenuItem('Маркировка стеноза'));
        self._menuItemMarkStenosis.setEnabled(false);


        var btnTools = new goog.ui.ToolbarMenuButton('Инструменты',toolsMenu);
        btnTools.setTooltip('Инструменты ФРК');
        t1.addChild(btnTools,true);

        // добавляем в меню Тесты
        var testsMenu = new goog.ui.Menu();

        var regMenu = new goog.ui.SubMenu("Регрессионные тесты");
        regMenu.addItem(new goog.ui.MenuItem('Почукаева-113'));
        regMenu.addItem(new goog.ui.MenuSeparator());
        regMenu.addItem(new goog.ui.MenuItem('debug test'));
        // regMenu.addItem(new goog.ui.MenuItem('Домаскин'));
        // regMenu.addItem(new goog.ui.MenuItem('Завьялова'));
        testsMenu.addItem(regMenu);

        testsMenu.addItem(this._menuItemLWSTests = new goog.ui.MenuItem('lws tests'));
        this._menuItemLWSTests.setEnabled(false);
        testsMenu.addItem(this._menuItemSvetz = new goog.ui.MenuItem('test load svetz-190'));
        this._menuItemSvetz.setEnabled(false);
        var btnTests = new goog.ui.ToolbarMenuButton('Тесты',testsMenu);
        btnTests.setTooltip('Тесты LWS');
        t1.addChild(btnTests,true);

        var btnAbout = new goog.ui.ToolbarMenuButton('О программе');
        btnAbout.setTooltip(this._about);
        //btnAbout.setEnabled(false);
        t1.addChild(btnAbout,true);        

        this._semafor= new goog.ui.ToolbarButton("🔴");
        this._semafor.setTooltip(`FFR сервер @${this._wsurl} не подключен`);
        t1.addChild(this._semafor,true);

        t1.render(this.mainwin);
        // прокси-функция, которая передает правильный контекст this в processAction()
        function processAction(e) {
            self.processToolbarAction_(e);
        }
        goog.events.listen(btnFile, 'action', processAction);
        goog.events.listen(btnTools, 'action', processAction);
        goog.events.listen(btnTests, 'action', processAction);

        var t1Element = t1.getElement();
        t1Element.className = "ffr-viewer-toolbar";
        t1Element.style.display="block";
        t1Element.removeAttribute("role");
        t1Element.removeAttribute("tabindex");

        return t1;
    }   // end createToolbar

    /**
     * FFR.Viewer.processToolbarAction_()
     * @param {goog.events.Event} e
     * @private
     */
    // e.target instanceof goog.ui.ToolbarToggleButton
    // e.target === btnInfo
    // e.target.getCaption() == "Инфо"
    // e.target.getContent() возвращает DOM <div class="V5-toolbar-button">
    //console.log('btnMenu.action event=');
    // e.type=="action" e.{altKey|ctrlKey|shiftKey|metaKey}
    // e.target=goog.ui.menuItem - значит доступны все его методы
    // e.target.getContent() === "Текст"

    processToolbarAction_(e)  {
        let cap = e.target.getCaption()
        const parentCap = e.target.getParent().parent_.content_   // химия
        console.log(cap);

        if(cap === 'Маркировка стеноза') {
            // это имитация маркировки стеноза с предварительной загрузкой state4
            const self = this;
            async function __markStenosisWithLoadState4() {

                console.log('Автоматический тест Почукаева-113')
                const _FFRsrv = FFR.viewer._FFRsrv
                const files = ['http://localhost/~germank/IM1']
                let loader = new AMI.VolumeLoader(/*container, CustomProgressBar*/);
                console.log(`загружаем DICOM ${files[0]}`)
                await loader.load(files)

                let series = loader.data[0].mergeSeries(loader.data)[0];
                console.log('прореживаем серию до 200 срезов' )
                let thinSeries = FFR.viewer.seriesLists.seriesThinOut(series,200);    // до 200 срезов
                //_seriesFingerprint(thinSeries); return
                console.log('добавляем серию в список и загружаем в QuadView' )
                FFR.viewer.seriesLists.addNewSeries(thinSeries);
                FFR.viewer._quadview.loadSeries(thinSeries);

                let state4 = await self._FFRsrv.sendRequestFSF("test load state4");
                //console.log('state4',state4)
                let aortaCube = new Cube(state4[0],'aorta'), 
                    v0Cube = new Cube(state4[1],'vessel0'), 
                    v1Cube = new Cube(state4[2],'vessel1');

                // почему-то в конце state4 находится null, поэтому length-1
                const vesselsCnt = (state4.length-1)/3 - 1;
                //console.log(`число сосудов в state4 ${vesselsCnt}`)
                let vesselInfo = [];
                for(var i=1; i<=vesselsCnt; i++) {
                    vesselInfo.push({name:state4[3*i], burificationPoints:state4[3*i+1].data, centerline:state4[3*i+2].data})
                }

                aortaCube.compress(); v0Cube.compress(); v1Cube.compress();      
                let state4Obj={stack:self._quadview._series.stack[0], 
                    aortaCube:aortaCube, 
                    vessels: [v0Cube,v1Cube], 
                    vesselInfo: vesselInfo,
                };

                self.processMarkStenosis(state4Obj)
            }
            
            __markStenosisWithLoadState4();

        }
        else if(cap.indexOf('Чтение DICOM каталога') !== -1 || (cap === 'Чтение DICOM файла')) {
          let inputFile = e.target.getElementByClass('inputfile-invisible');
          const self = this;
          let onChange = function(e) {  // не может быть лямбда функцией потому, что надо делать removeEventListener
            inputFile.removeEventListener('change', onChange);
            handler__loadDICOMFileOrDir(e.target.files);
            // https://stackoverflow.com/questions/19643265/second-use-of-input-file-doesnt-trigger-onchange-anymore
            e.target.value = ''
          }
          inputFile.addEventListener('change', onChange);
        }

        /// Меню Tесты
        ///
        else if(cap === 'lws tests') {
          //window.location.reload()
            this._FFRsrv = new FFR__protocol(this._wsurl);       
            this._FFRsrv.addEventListener('ready', async (event) => {
                console.log('ready');
                const self = this;
                async function __2() {
                    let result = await self._FFRsrv.sendRequestReturningArrayBuffer("test lws");
                    //console.log('result',result)
                    let obj = FSF.parseBuffer(result); 
                    //console.log('obj',obj);
                    const cube = new Cube(obj[0],'zero cube')
                    cube.nonzerostats();
                    console.log(cube)
                }

                __2();
            });
        }
        else if(cap === 'test load svetz-190') {
            this._FFRsrv = new FFR__protocol(this._wsurl);       
            this._FFRsrv.addEventListener('ready', async (event) => {
                console.log('ready');
                const self = this;
                async function __1() {
                    // let data0 = await self._FFRsrv.sendRequestReturningArrayBuffer("test Float32Array long");
                    // // console.log('data',data0); return;
                    // let data1 = FSF.parseBuffer(data0);
                    // console.log('data',data1); return;

                    let data = await self._FFRsrv.sendRequestReturningArrayBuffer("test load svetz-190");
                    let shvetz190 = FSF.parseBuffer(data);
                    // console.log('shvetz190',shvetz190); return;
                    let frame_i_data = await self._FFRsrv.sendRequestReturningArrayBuffer("test load svetz-frame-i");
                    let frame_i = FSF.parseBuffer(frame_i_data)[0];
                    let frame_k_data = await self._FFRsrv.sendRequestReturningArrayBuffer("test load svetz-frame-k");
                    let frame_k = FSF.parseBuffer(frame_k_data)[0];
                    // console.log('frame_k',frame_k); return;
                    let shvetzCube = new Cube(shvetz190[0],'shvetz')
                    // функция shvetz определена в shvetz.js и используется только для тестов
                    // shvetz(shvetzCube,frame_i,frame_k)
                }

                __1();
            });
     
        }
        else if(parentCap === 'Регрессионные тесты')
            RGT.processEvent(e)
    }

    /**
     * FFR.Viewer.loadAortaMask(file) параметр или File или Blob
     * @param {!File|Blob} file
     * @return {Promise}
     */
    loadAortaMask(file) {
      console.log('loadAortaMask()',file);
      return new Promise(function(resolve, reject) {
        let reader = new FileReader();
        reader.onload = function() {
          console.log('res=',reader.result);
          let mask = new Uint8Array(/** @type {ArrayBuffer} */ (reader.result));
          //console.log('mask.byteLength=',mask.byteLength);  // .length - число элементов в массиве
          let firstNz=-1, lastNz=-1, cntNz=0;
          for(let i=0; i<mask.length; i++) {
            if(mask[i] == 0) continue;
            if(firstNz == -1) firstNz = i;
            if(lastNz < i) lastNz = i;
            cntNz++;
          }
          console.log('firstNz=',firstNz,'lastNz=',lastNz,'cntNz=',cntNz);

          // initAMI();
          
          let width=512, height=512, depth = 449;
          console.log('calling MarchingCubes',performance.now());
          let aortaCube = new Cube({dims:[512,512,113], data:mask}, 'AotraMask');
          let aortaGeometry = aortaCube.geometry();
          this.aortaPreviewGeometry = aortaGeometry;
          console.log('aortaPreviewGeometry', aortaGeometry)
          resolve(aortaGeometry);
        };
        reader.onerror = function() {
            console.log(reader.error);
            reject(reader.error);
        };

        reader.readAsArrayBuffer(file);
      });      
    }

    showSemafor(green,version) {
        const self = this
        if(!green) {
            self._semafor.setCaption('🔴');
            self._semafor.setTooltip('FFR сервер закрыл подключение');
            alert('FFR сервер закрыл подключение');
            self._menuItemMarkStenosis.setEnabled(false);
        }  
        else {
            self._semafor.setCaption('🟢');                        
            //let version = await this._FFRsrv.version();
            self._semafor.setTooltip('FFR сервер (версия '+version+') готов @'+this._wsurl);
            //self._menuItemMarkStenosis.setEnabled(true);            
        }
    }

    /**
     * FFR.Viewer.showInfoMessage(message) 
     * @param {!string} message
     */
    showInfoMessage(message) {
      alert(message);   // потом надо будет оформить как модальный диалог
    }

    set FFRsrv(ffrsrv) {
        this._FFRsrv = ffrsrv;
    }

    set ffrDriver(driver) {
        this._ffrDriver = driver;
    }

}

function logElementSize(el) {
  //var el = this.dialog.dialog.getContentElement()
  var borderBoxSize = goog.style.getBorderBoxSize(el);
  var size = goog.style.getSize(el);
  var computedBoxSizing = goog.style.getComputedBoxSizing(el);
  var visibleRectForElement = goog.style.getVisibleRectForElement(el);
  var bounds = goog.style.getBounds(el);
  var contentBoxSize = goog.style.getContentBoxSize(el);

  console.log('Dialog12 contentElement borderBoxSize=',borderBoxSize)
  console.log('Dialog12 contentElement size=',size)
  console.log('Dialog12 contentElement computedBoxSizing=',computedBoxSizing);
  console.log('Dialog12 contentElement visibleRectForElement=',visibleRectForElement);
  console.log('Dialog12 contentElement bounds=',bounds);
  console.log('Dialog12 contentElement borderBoxSize=',borderBoxSize);
  console.log('Dialog12 contentElement contentBoxSize=',contentBoxSize);
}
